Moving AMC back outside docker

OmarIthawi opened the pull request:
One more desperate attempt to make devstack works on Mac.

### So, what's the point of this?
Before Dockerizing AMC (#21) we haven't had too much issues running the devstack on @grozdanowski's device. Now we do.

It's either the Frontend, or the backend. Now I'm starting with the easier part, the frontent. Also the AMC frontend run over 30+ watcher processes, which is just too much!

```
$ docker ps -q | xargs  docker stats --no-stream
CONTAINER ID        NAME                      CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
ed7098c038a4        tahoe.devstack.amc        4.49%               104MiB / 15.42GiB     0.66%               34.8kB / 164kB      6.91MB / 590kB      4
d8a103ddf099        edx.devstack.studio       0.02%               503MiB / 15.42GiB     3.19%               50.4kB / 2.99kB     7.05MB / 983kB      4
c9b36e974b9d        edx.devstack.lms          0.02%               525.8MiB / 15.42GiB   3.33%               50.9kB / 2.99kB     49.5MB / 1.16MB     4
8cd6c6b0040f        edx.devstack.mongo        1.04%               30.61MiB / 15.42GiB   0.19%               11.1kB / 0B         4.65MB / 1.62MB     15
3e5fe1005465        tahoe.devstack.redis      0.11%               2.547MiB / 15.42GiB   0.02%               10.4kB / 0B         8.55MB / 0B         3
9a824ebc97ba        edx.devstack.memcached    0.00%               1.777MiB / 15.42GiB   0.01%               9.33kB / 0B         2.2MB / 0B          6
9df661c638b8        edx.devstack.devpi        0.17%               51.9MiB / 15.42GiB    0.33%               10.2kB / 0B         12.5MB / 8.19kB     52
fda67c4d23f2        edx.devstack.mysql        0.09%               463.6MiB / 15.42GiB   2.94%               15.2kB / 83.3kB     10MB / 90.1kB       23
e8612e804815        tahoe.devstack.postgres   0.00%               9.734MiB / 15.42GiB   0.06%               17.8kB / 14.2kB     12.7MB / 369kB      7
```

None of the above containers (except for MySQL) has more processes (PIDS) more than the AMC Frontend:

```
7ae41fd52620        tahoe.devstack.amc-frontend   0.00%               330.9MiB / 15.42GiB   2.10%               2.68MB / 73.2kB     788MB / 1.14MB      39
```

OmarIthawi commented:
@ahmedaljazzar fyi. But don't try this yet.

johnbaldwin reviewed (approved):
I'mm willing to give this a try. Thanks @OmarIthawi !

OmarIthawi commented:
Thanks @johnbaldwin, I'm still cleaning this up. Marking as `(WIP)` for now.

OmarIthawi commented:
@ahmedaljazzar checkout the PR branch and hit:

```
$ make tahoe.clean-provision
```

**WARNING:** This will destroy all the Docker volumes you have.

iamjazzar commented:
I went through those errors:
- Hard coded `~/work/tahoe-hawthorn`
- A lot of version conflict in the way
```
VersionConflict: (edx-i18n-tools 0.4.6 (/edx/app/edxapp/venvs/edxapp/lib/python2.7/site-packages), Requirement.parse('edx-i18n-tools==0.4.5'))
```
- The command finally failed at some npm error.

```
../../nan/nan_converters_43_inl.h:22:1: warning: 'ToBoolean' is deprecated: ToBoolean can never throw. Use Local version. [-Wdeprecated-declarations]
.
.
.
10 warnings and 8 errors generated.
make[2]: *** [Release/obj.target/binding/src/binding.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/ajazzar/projects/appsembler/tahoe-hawthorn/amc/frontend/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:256:12)
gyp ERR! System Darwin 18.6.0
gyp ERR! command "/usr/local/Cellar/node/12.1.0/bin/node" "/Users/ajazzar/projects/appsembler/tahoe-hawthorn/amc/frontend/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /Users/ajazzar/projects/appsembler/tahoe-hawthorn/amc/frontend/node_modules/node-sass
gyp ERR! node -v v12.1.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
Build failed with error code: 1
npm WARN @fortawesome/react-fontawesome@0.1.4 requires a peer of react@16.x but none is installed. You must install peer dependencies yourself.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 install: `node install`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-sass@4.11.0 postinstall: `node scripts/build.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-sass@4.11.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ajazzar/.npm/_logs/2019-07-18T11_26_54_955Z-debug.log
make[1]: *** [amc.provision] Error 1
make: *** [dev.provision] Error 2
```

OmarIthawi commented:
Thanks for the review @ahmedaljazzar! I've addressed all of the issues you've faced:

> I went through those errors:
> 
> * Hard coded `~/work/tahoe-hawthorn`

Oops! I forgot to use the proper variable. Done.

> * A lot of version conflict in the way
> 
> ```
> VersionConflict: (edx-i18n-tools 0.4.6 (/edx/app/edxapp/venvs/edxapp/lib/python2.7/site-packages), Requirement.parse('edx-i18n-tools==0.4.5'))
> ```

The `appsembler/edxapp` Docker image was a couple of days old on your device. I added a `docker-compose pull` step to make sure this problem goes away.

> * The command finally failed at some npm error.
> 
> ```
> ../../nan/nan_converters_43_inl.h:22:1: warning: 'ToBoolean' is deprecated: ToBoolean can never throw. Use Local version. [-Wdeprecated-declarations]
> .
> gyp ERR! node -v v12.1.0
> gyp ERR! node-gyp -v v3.8.0

Node.js is too recent (v12) we need v10 for the AMC. The Makefile now checks for that and suggest a solution.

johnbaldwin reviewed (approved):
@OmarIthawi I haven't been able to test this yet, and there's significant changes that I'm still getting up to speed on in devstack + docker.

I may have time this weekend to test it on my machine if I get done with the API docs. I'm willing to approve and if we run into problems, then fix and do another PR

OmarIthawi commented:
> @OmarIthawi I haven't been able to test this yet, and there's significant changes that I'm still getting up to speed on in devstack + docker.
> 
> I may have time this weekend to test it on my machine if I get done with the API docs. I'm willing to approve and if we run into problems, then fix and do another PR

Thanks @johnbaldwin! However, I'm going to re-request your review and will only merge after two MacOSX developers test the changes and approve it.

We've used light reviews a lot, but for this project it's not working well and therefore I'm reverting to in-depth reviews.

You don't have to check the changes, a full reset to the devstack to see if the `getcwd` issue is gone is enough for this specific PR.

iamjazzar commented on tahoe.mk:
```diff
@@ -40,16 +32,65 @@ tahoe.restart:  ## Restarts both of LMS and Studio python processes while keepin
+ifneq ($(shell node --version | grep -o '^v10\.'), v10.)
+	echo "AMC needs Node.js v10, no more no less."
+	echo "You DO NOT have Node.js v10 installed. Please install it."
+	echo "Try running $ tahoe.macosx.install-nodejs-v10 on MacOSX."
```
echo "Try running $ **_make_** tahoe.macosx.in ..."

iamjazzar commented:
@OmarIthawi Thanks for the fixes. The provisioning process now freezes at `make amc.start.frontend`
```
Time: 8653ms
                                                     Asset     Size  Chunks             Chunk Names
                      ae440fff5bf947898b17a506933f9acb.jpg   117 kB          [emitted]  
                      3feb378ad4bc8cb8b206d745de5ef995.jpg   227 kB          [emitted]  
                      eab2bf015cccb03d12ea7a029c0b3574.jpg   165 kB          [emitted]  
                      00de48b8d04dd2e922c0b4cca4af803e.jpg   277 kB          [emitted]  
                      797c184d81d439f1bc25cd71a8efe4ca.jpg   432 kB          [emitted]  
                      d60187eca115fe4f264f9c53ceef756c.jpg   207 kB          [emitted]  
                      31c6f7e07841bb091b01f898afce378f.jpg   222 kB          [emitted]  
                      3ee42f95b8c53d299ceb2ac0643f99ca.png  35.7 kB          [emitted]  
src/static/img/design-templates/app_store_badge_135x40.svg  12.2 kB          [emitted]  
                      9f4b4147e434e41ba3c4d34a8a9586d3.jpg   137 kB          [emitted]  
                      ebb6db433f7c7bc8766111baee662b7f.jpg   153 kB          [emitted]  
                      65d62d67d1876a91e03b0c211b0b7b9b.jpg  76.2 kB          [emitted]  
                      0ea8993d4f44210a49513a2f9b678459.jpg   196 kB          [emitted]  
                      723471d465f075ef67c22b4f198b8435.jpg  20.7 kB          [emitted]  
                      df6f3f80bb89236c39ea7b6821ac5f77.jpg   255 kB          [emitted]  
                      70ff0a7f6c984510c3c510212baf3b3f.jpg   201 kB          [emitted]  
                      92499caf3b57e832e20de0a26fd40e75.jpg   159 kB          [emitted]  
                      02e6863440035138dc8c4bb5f78bf2c5.jpg   318 kB          [emitted]  
                      ccad9211d05be8dde50bc6664edda8d4.jpg   181 kB          [emitted]  
                      8eadf9b90460a824e81fddb73def8ac7.jpg  86.8 kB          [emitted]  
                               app.9937281d50359a7b6767.js  9.36 MB       0  [emitted]  app
                            public.9937281d50359a7b6767.js  26.3 kB       1  [emitted]  public
                            vendor.9937281d50359a7b6767.js  2.05 MB       2  [emitted]  vendor
                           app.9937281d50359a7b6767.js.map  10.3 MB       0  [emitted]  app
                        public.9937281d50359a7b6767.js.map  27.8 kB       1  [emitted]  public
                        vendor.9937281d50359a7b6767.js.map  2.47 MB       2  [emitted]  vendor
webpack: Compiled successfully.
```

OmarIthawi commented:
@ahmedaljazzar that's a feature :wink: it's a gift for you for having a successful provision! Please consider giving this PR a :heavy_check_mark:

iamjazzar commented:
Well, I don't really think freezing while provisioning is a good idea. That was totally counter intuitive. I also tried stopping the server (`make stop`) and running it again (`make tahoe.up`) but I got the a blank site on AMC with the following errors:

```
[Error] Failed to load resource: Could not connect to the server. (vendor.9937281d50359a7b6767.js, line 0) http://0.0.0.0:13000/static/webpack/vendor.9937281d50359a7b6767.js
[Error] Failed to load resource: Could not connect to the server. (app.9937281d50359a7b6767.js, line 0) http://0.0.0.0:13000/static/webpack/app.9937281d50359a7b6767.js
[Error] Failed to load resource: the server responded with a status of 404 (Not Found) (None.js, line 0) http://js.hs-scripts.com/None.js
```

OmarIthawi commented on tahoe.mk:
```diff
@@ -40,16 +32,65 @@ tahoe.restart:  ## Restarts both of LMS and Studio python processes while keepin
+		docker build -f Dockerfile . -t gcr.io/appsembler-tahoe-0/amc
+
+	make dev.provision
+	make amc.start.frontend
```
```suggestion
	@echo "Provisioning Tahoe is now complete. Please run the following command to start the AMC frontend (outside the Docker devstack: '$ make amc.start.frontend')"
```

@ahmedaljazzar this is what you're looking to change.

OmarIthawi replied:
> Well, I don't really think freezing while provisioning is a good idea. That was totally counter intuitive. I also tried stopping the server (`make stop`) and running it again (`make tahoe.up`) but I got the a blank site on AMC with the following errors:
> 
> ```
> [Error] Failed to load resource: Could not connect to the server. (vendor.9937281d50359a7b6767.js, line 0) http://0.0.0.0:13000/static/webpack/vendor.9937281d50359a7b6767.js
> [Error] Failed to load resource: Could not connect to the server. (app.9937281d50359a7b6767.js, line 0) http://0.0.0.0:13000/static/webpack/app.9937281d50359a7b6767.js
> [Error] Failed to load resource: the server responded with a status of 404 (Not Found) (None.js, line 0) http://js.hs-scripts.com/None.js
> ```

@ahmedaljazzar I think there's a misunderstanding here, so I would like to clarify few things:

 - The whole point of this pull request is to move out the AMC Frontend outside the docker devstack.
 - Moving the AMC frontend outside the devstack means that it won't be included in the `$ docker-compose up` (or `$ make tahoe.up`), so a seperate shell running `$ make amc.start.frontend` is needed otherwise you'd get the 404 errors
 - I decided to add this as an extra step for my own convenient so I don't have to maintain two shells when I reset the devstack. I have removed this step (see https://github.com/appsembler/devstack/pull/30#discussion_r305830927), but you still have to run it on a separate terminal otherwise you'd get the 404 errors
- This is still a huge `(WIP)` PR and the `$ make amc.start.frontend` is already known to folks like Maxi and Matej so I skipped on editing the documentation till we test the PR and consider it :heavy_check_mark:

iamjazzar reviewed (approved):
Everything LGTM here.  I tested the PR on a low-specs machine and I couldn't find any issue on my side.

> **Processor** 2.3 GHz Intel Core i5
> **Memory** 8 GB 2133 MHz LPDDR3
> **Graphics** Intel Iris Plus Graphics 655 1536 MB

OmarIthawi committed:
Refactor: Moving AMC back outside docker
